### PR TITLE
[LOOP-26] Upload loop dosing decision data to Tidepool backend

### PR DESCRIPTION
**Pull request by darinkrauss**
- https://tidepool.atlassian.net/browse/LOOP-26
- Add Codable to StoredGlucoseSample to persist in dosing decision
- Make bolusUnits optional in AutomaticDoseRecommendation
- Unify DosingDecisionStore in LoopKit from Loop
- Update StoredDosingDecision properties
- Add time zone to StoredSettings properties
- Refactor how errors and warnings are captured in StoredDosingDecision
- Update tests

### CODE REVIEW

File: LoopKit/DosingDecisionStore.swift [hunk 1]
```diff
@@ -84,12 +86,29 @@ public class DosingDecisionStore {
...
+    private static var encoder: PropertyListEncoder = {
+        let encoder = PropertyListEncoder()
+        encoder.outputFormat = .binary
+        return encoder
+    }()
+
+    private func encodeDosingDecision(_ dosingDecision: StoredDosingDecision) -> Data? {
+        do {
+            return try Self.encoder.encode(dosingDecision)
+        } catch let error {
```
Comment (rickpasetto) [inline LoopKit/DosingDecisionStore.swift:98]: [nit] I don't think you need the `let error` anymore in Swift.  `error` automatically is available

File: LoopKit/DosingDecisionStore.swift [hunk 2]
```diff
@@ -173,70 +196,106 @@ public struct StoredDosingDecisionData {
     }
 }
 
+public typealias HistoricalGlucoseValue = PredictedGlucoseValue
+
 public struct StoredDosingDecision {
```
Comment (rickpasetto) [inline LoopKit/DosingDecisionStore.swift:201]: [nit] consider moving this struct (and it's dependent friend structs) to its own file (just one file is fine)

**Review by rickpasetto — APPROVED** [review]
A few nits, other than that LGTM

File: LoopKitTests/SettingsStoreTests.swift [hunk 3]
```diff
@@ -1004,12 +1011,13 @@ extension StoredSettings: Equatable {
 
 fileprivate extension StoredSettings {
     static var test: StoredSettings {
+        let timeZone = TimeZone(identifier: "America/Los_Angeles")!
```
Comment (ps2) [inline LoopKitTests/SettingsStoreTests.swift]: Does this represent the timezone that the device is currently in? Or the schedules? The schedules do not get normal timezones applied; they have fixed dst offset timezones; using timezones that participate in DST can cause issues when tests run in different parts of the year. Maybe we should update the name to make which type it is clear?

**Review by ps2 — COMMENTED** [review]
LGTM. Just a question about the new timezone field.

File: LoopKitTests/SettingsStoreTests.swift (hunk 3 above)
Comment (ps2) [reply LoopKitTests/SettingsStoreTests.swift]: Which one does it represent?

Comment (darinkrauss) [reply LoopKitTests/SettingsStoreTests.swift]: This particular time zone field represents the time zone the controller is in when the StoredSettings were created. This has been renamed to `controllerTimeZone` to be more descriptive.

**Pull request merged**